### DISCUSSION TITLE
[IDEA] Make "Listeners" not regular users

### DISCUSSION THREAD

**Discussion by 38github** [jamulussoftware/jamulus#3283] [Ideas]
Maybe it is also possible to make them use less resources like less bitrate, larger framesize, etc and lower priority to minimize CPU and network usage. Is this something that is of interest to implement?

Maybe also make regular users that are muted for a certain amount of time not able to be a regular user and automatically becomes a "listener user" allowing opening user slot on the server for active participants.

**Comment by mcfnord** [comment]
I imagine a pure listener doesn't need all of the UI that audible participants need.

**Reply by dcorson-ticino-com** [reply]
But adding special cases to the protocol doesn't make things simpler.
I don't think that Jamulus is a resource hog, we have seen groups of around 100 users playing together.
If there are a couple listeners in the lounge it makes no real difference.
No action required IMHO.

Adding non-compressed audio would make a big difference 'though.  That would be something to work on.

**Reply by 38github** [reply]
Ther are servers that only allow a small amount of users and I know that some of them do this because of limited resources and also because of being in a VM that according to some users increases CPU because of the low latency network traffic.

**Reply by dcorson-ticino-com** [reply]
In that case I would say that HW is cheap.  During Covid I was running a server on a Linux SBC that cost about $100 and had no problems with 20 users.
The reduction in server CPU power needed by eliminating the sound input for a listener is minimal.  Listeners still would need to run the full protocol, mixing, compression, etc.

**Reply by 38github** [reply]
I understand.

**Comment by mcfnord** [comment]
The server might not benefit from listener-only connections, but a listen-only client in a browser that lets you set levels would be popular.

**Reply by dcorson-ticino-com** [reply]
We have that already, just press "mute myself".

**Reply by mcfnord** [reply]
The download is friction. I can't do it on my work PC, for example. I'll try it again on my Android phone.

**Reply by Benjamin-Y-Wang** [reply]
I have implemented a prototype implementing Jamulus protocol some time ago and have been testing it on my own. The demo video can be found here: https://www.youtube.com/watch?v=g4ERzniMBGo 

It has some advantages for a pure Jamulus listener:
1. Doesn't require a standalone audio interface, onboard sound device works fine, no ASIO required on Windows either as low-latency is becoming less important if you are only listening. Also since no ASIO involved, you don't need to specially bridge audio to WDM if you are live streaming with it, softwares like OBS can pick the application audio directly.
2. Removed almost all configurations in Jamulus software, basically you can only set your own name, your role will be locked as "Listener", so it's much easier to use especially for new players.
3. Hiding yourself (and other similar users using the same software) from your own mixing panel, so only real players are displayed to make UI as tidy as posible.
4. It provides a more mixer-like look & feel, added the dB markers to the mixing slider and set level to 0 dB initially, this will provide +10dB room to boost up quiet channels.
5. Added a main volume control on top of mixed audio, this will make it easier collaborating with other applications, such as you are live streaming and runs background music in your audio player, the volume are easier to balance between these applications.

Meanwhile it keeps remaining Jamulus features as many as possible, channel gain/pan/mute/solo, and level meter are all implemented. The chat function is also there, you can find it at end of demo video.

There are certainly many limitations keep it only as a prototype instead of usable software:
1. UI is ugly, I'm not a frontend developer so skill is limited.
2. Also not build it as a desktop application, but a mini local web server communicating with Jamulus server, and exposing web UI to user. This is also due to my limited skillset, I'm a server side software engineer and used golang to develop.
3. Windows only, theoriotially golang can cross some platform but Win is the only version I have built so far, no mobile version either.
4. No localization, UI text is in Chinese so far.

This is not a mature product but just like a practice of me learning Jamulus protocal. Also my work is inspired by https://github.com/dtinth/gojam, also a golang client but works on server side. I have read through the good work from dtinth and tried to implement my own version of it.

Anyway I guess this could be considered as a good enough example to prove such "listen only" idea is feasible, although I'm still far away from implement a satisfied version yet.

**Reply by ann0see** [reply]
Listen only web clients exist already. @mcfnord you can put your repos link here.
